src/bisulfite_scoring.hpp

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2016-2019, Knut Reinert and Freie Universität Berlin
+// All rights reserved.
+//
+// This file is part of Lambda.
+//
+// Lambda is Free Software: you can redistribute it and/or modify it
+// under the terms found in the LICENSE[.md|.rst] file distributed
+// together with this file.
+//
+// Lambda is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+//
+// ==========================================================================
+// Scoring schemes for bisulfite converted data on dna5 alphabet
+// ==========================================================================
+
+#pragma once
+
+#include <range/v3/algorithm/copy.hpp>
+
+#include <seqan3/alignment/scoring/scoring_scheme_base.hpp>
+#include <seqan3/alphabet/nucleotide/dna5.hpp>
+#include <seqan3/range/shortcuts.hpp>
+#include <seqan3/std/algorithm>
+
+namespace seqan3
+{
+
+template <arithmetic score_type = int8_t>
+class bisulfite_scoring_scheme : public scoring_scheme_base<bisulfite_scoring_scheme<score_type>, dna5, score_type>
+{
+private:
+    using base_t = scoring_scheme_base<bisulfite_scoring_scheme<score_type>, dna5, score_type>;
+    using base_t::matrix;
+public:
+    using base_t::base_t;
+    using typename base_t::matrix_type;
+    using matrix_size_type = std::remove_const_t<decltype(alphabet_size<dna5>)>;
+    static constexpr matrix_size_type matrix_size = alphabet_size<dna5>;
+
+    constexpr bisulfite_scoring_scheme() noexcept {}
+
+    template <arithmetic score_arg_t>
+    constexpr bisulfite_scoring_scheme(match_score<score_arg_t> const ms, mismatch_score<score_arg_t> const mms)
+    {
+        set_bisulfite_scheme(ms, mms);
+    }
+
+    constexpr bisulfite_scoring_scheme(matrix_type const & _matrix) noexcept {}
+
+    template <arithmetic score_arg_t>
+    constexpr void set_bisulfite_scheme(match_score<score_arg_t> const ms, mismatch_score<score_arg_t> const mms)
+    {
+        std::conditional_t<std::integral<score_type>, int64_t, double> i_ms = static_cast<score_arg_t>(ms);
+        std::conditional_t<std::integral<score_type>, int64_t, double> i_mms = static_cast<score_arg_t>(mms);
+        if ((i_ms  < std::numeric_limits<score_type>::lowest() || i_ms  > std::numeric_limits<score_type>::max()) ||
+            (i_mms < std::numeric_limits<score_type>::lowest() || i_mms > std::numeric_limits<score_type>::max()))
+        {
+            throw std::invalid_argument{"You passed a score value to set_bisulfite_scheme that is out of range of the "
+                                        "scoring scheme's underlying type. Define your scoring scheme with a larger "
+                                        "template parameter or down-cast you score value beforehand to prevent "
+                                        "this exception."};
+        }
+
+        // Assume query is horizontal sequence
```

Comment:
I would maybe override the score function from `scoring_scheme_base` to ensure that exactly one parameter is of type `dna3bs` and the other isn't? Seems like it is easy to get wrong otherwise...